doc/neps/nep-0054-simd-cpp-highway.rst

```diff
@@ -0,0 +1,480 @@
+
+1. If it cross-compiles with Clang and can be tested via standard QEMU, it can
+   go into Highway's CI.
+2. If it cross-compiles via clang/gcc and can be tested with a new QEMU
+   (possibly with extra flags), then it can be support via manual testing
+   before each Highway release.
+3. Existing targets will remain supported as long as they compile/run in QEMU.
+
+Highway is not subject to Google's "no longer supported" strategy (or, as
+written in its README, *This is not an officially supported Google product*).
+That is not a bad thing; it means that it is less likely to go unsupported due
+to a Google business decision about the project. Quite a few well-known open
+source projects under the ``google`` GitHub org state this, e.g. `JAX`_ and
+`tcmalloc`_.
+
+
+Supported features/targets
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Both frameworks support a large set of platforms and SIMD instruction sets,
+as well as generic scalar/fallback versions. The main differences right now are:
+
+- NumPy supports IBM Z-system (s390x, VX/VXE/VXE2) while Highway does not,
+- Highway supports ARM SVE/SVE2 and RISC-V RVV (sizeless instructions), while
+  NumPy does not.
+
+  - The groundwork for sizeless SIMD support in NumPy has been done in
+    `gh-21057`_, however SVE/SVE2 and RISC-V are not yet implemented there.
+
+Either of the above is "just work" - completing Highway will be less work than
+completing NumPy, but both are doable - and hence this should probably not be a
+deciding factor in the decision.
+
+There is also a difference in the granularity of instruction set groups: NumPy
+supports a more granular set of architectures than Highway. See the list of
+targets for Highway `here <https://github.com/google/highway/#targets>`__
+(it's roughly per CPU family) and for NumPy
+`here <https://numpy.org/doc/1.25/reference/simd/build-options.html#supported-features>`__
+(roughly per SIMD instruction set). Hence with Highway we'd lose some
+granularity - but that is probably fine, we don't really need this level of
+granularity, and there isn't much evidence that users explicitly play with this
+to squeeze out the last bit of performance for their own CPU.
+
+
+Compilation strategy for multiple targets and runtime dispatching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Highway compiles once while using preprocessing tricks to generate multiple
+stanzas for each CPU feature within the same compilation unit (see the
+``foreach_target.h`` usage and dynamic dispatch docs for how that is done).
+Universal Intrinsics generate multiple compilation units, one for each CPU
+feature group, and compiles multiple times, linking them all together (with
+different names) for runtime dispatch. The Highway technique may not work
+reliably on MSVC, the Universal Intrinsic technique does work on MSVC.
+
+Which one is more robust? The experts disagree. Jan thinks that the Highway
```

Comment:
The "build systems come and go" argument is a good one - I like it. Also, we have scenarios like Bazel users (like Google) pulling in NumPy and rewriting the build system for their monorepo needs. That doesn't have to be a primary concern for us, but indeed it's nicer if there was less dependence on build system features.

Reply:
Makes sense. I've also learned that the hard way. We started with Bazel, then came many users who wanted CMake, then  Chrome with their GN.

Reply:
From my perspective, your hypothesis here is not accurate. First, the SIMD frameworks are unstable and constantly changing, being replaced by newer ones more quickly than any build system. For example, as far as I know, Highway primarily exists to replace these frameworks that don't support scalable SIMD extensions. Am I correct? Additionally, does it provide a portable dynamic solution? So, in the future, projects may consider adopting lightweight JIT solutions, which is something we should use one day. This would eventually lead us to replace Highway, sooner or later. In general, nothing is static in this universe except CMake :).

However, I have proven to you that it's possible to support scalable SIMD extensions for frameworks that use object classes and operators by relying on the optimizer. Therefore, it would have been better to support a framework like xsimd or std instead of creating a new one.

Regarding the portable dynamic dispatching solution, we did support a similar solution based on compiler target attributes, which is similar to the current approach used by Highway. So, this wasn't new in NumPy.

The reason for replacing it was mainly due to compiler bugs, the difficulty in managing, debugging, and supporting MSVC. Even with compiler target attributes, we had to test each attribute against the compiler to ensure it was running correctly. Otherwise, we had to track each compiler version or subversion to mark it as a broken platform similar to what Highway do, which is not a practical approach for projects like NumPy.

However, even with the use of Highway's dynamic dispatching, we still need to deal with the build system. For example,  provide build options to allow the user to disable or enable certain features, raise the baseline features, or support native features. we should also provide a build log to track the enabled features and generated kernels.

Reply:
Is there a "takeaway" from this discussion that should be added to the NEP, or is the text accurate here?

Reply:
> Yes, that is possible. Though I hear about database projects moving away from these again because they are difficult to inspect/debug.

[asmjit](https://github.com/asmjit/asmjit) is a promising project, we may use it one day to generate our ufunc loops from python with SIMD support and with ability to support loop fusion. [QuestDB](https://github.com/questdb/questdb) is already use it.

Reply:
>  The one remaining downside is that numpy is then quite fundamentally tied to Meson, and it will be difficult to switch to or even support other build systems.

I agree but this also gives the ability to support C language which is important to deal with our legacy C code especially when it comes to cache and tracing the dispatches calls.

Reply:
I will add this

Both agree the meson build system allows specifying object link order,
which produces more consistent builds. However that does tie NumPy to meson.

Matti thinks the current build strategy is working well for NumPy and the
advantages of changing the build and runtime dispatch, with possible unknown
instabilities outweighs the advantages that fully adopting Highway may bring.